src/site/topics/queues/linked-queue.rst

```diff
@@ -2,270 +2,231 @@
-        if (isEmpty()) {
-            rear = null;
-        }
-        return returnElement;
-    }
-
-    @Override
-    public T first() {
-        if (isEmpty()) {
-            throw new NoSuchElementException("First from an empty queue.");
-        }
-        return front.getData();
-    }
+
+
+``dequeue`` & ``first``
+-----------------------
+
+.. literalinclude:: /../main/java/LinkedQueue.java
+    :language: java
+    :lineno-match:
+    :lines: 42-63
+
+
 
 * Like ``LinkedStack`` and ``ArrayStack``, trying to access something from the empty queue throws an exception
+* ``dequeue`` does a *remove/delete from the front of a linked structure*
 
-* Notice in ``dequeue`` that, if it's not empty, we just *remove/delete from the front of a linked structure*
-    * This was the same as ``pop`` in the ``LinkedStack``
+* Notice the edge case of a ``dequeue`` resulting in an empty queue
 
-* In addition to being empty, the only other edge case we need to watch out for is if the ``dequeue`` makes the queue empty
-* If this happens, we must set ``front`` to ``null``
-    * This is actually taken care of already since ``front``'s next would be ``null``, and saying ``front = front.getNext()`` makes ``front`` ``null``
+    * The update to the ``front`` will not be a problem
 
-* We should also set ``rear`` to ``null``
-* This helps with garbage collection and keeping the state of the queue *correct*
-* If we don't do this, ``rear`` will continue to point to a node that should not be in the queue anymore
-    * Our current implementation will handle this scenario fine as an enqueue on an empty queue sets both ``front`` and ``rear`` to ``null``
-    * But imagine also having a messed up count and ``enqueuing`` after this without setting ``rear`` to ``null``
-    * If we ``enqueue`` in this case, we might end up saying ``rear.setNext(toEnqueue)``
-    * What would that mean?
-    * What would that look like?
+        * ``front = front.getNext()`` will set ``front`` to ``null`` since ``front.getNext()`` in this situation would return ``null``
 
+    * The trouble is with ``rear`` --- typically ``rear`` is left alone with a ``dequeue``
+    * However, if the ``dequeue`` resulted in an empty queue, ``rear`` would be left referencing the ``Node`` that was just removed
+    * The solution is to simply set it to ``null`` if the stack is empty after the dequeue
 
-* What is the computational complexity of a ``dequeue``?
 
-Variations
-==========
+.. note::
+
+    Although not true in general, with the provided implementation of the ``LinkedQueue``, missing this edge case in the
+    ``dequeue`` would not actually cause a problem since the ``enqueue`` was written such that it also checks for the
+    empty case.
+
+    However, ignoring this edge case in the ``dequeue`` because it is caught in the ``enqueue`` only further couples the
+    methods together. It is better to have independent functional units be correct. Further, removing *dangling* edge
```

Comment:
Significantly updated the note to be clearer. 

I feel like this wording, "It is better to have independent functional units be correct", could be improved slightly, but I just can't get the words out.

Reply:
Although true, it still makes them that much more coupled.

Reply:
Correct. That's what happens when logic and data are combined into a mutable entity. There's no getting around it when you have mutable state, even strategies like sentinel nodes still introduce some form of coupling (coupling to the node, rather than another method).

The way around it is to make the queue immutable (with a `head` + `tail`), but, let's be real, an immutable queue is also annoying to use in typical cases (like breadth first search). 

The note here is essentially, test the interface, not the implementation. Find all the states that your class can be in and assert the behaviour is as you expect in those states. It just so happens that the interface of a mutable collection is coupled out the whazoo.

Reply:
I think the best solution is to just cut the note back down as it's starting to get too verbose for the initial point of _yeah, sure, this implementation is fine, but watch for the edge cases_

Reply:
there --- simplified.